phoenix-core/src/test/java/org/apache/phoenix/cache/ServerMetadataCacheTest.java

```diff
@@ -419,16 +423,291 @@ public void testInvalidateCacheForBaseTableWithUpdateIndexStatement() throws Exc
+            createTable(conn1, tableName, NEVER);
+            upsert(conn1, tableName);
+
+            // Instrument ServerMetadataCache to throw a SQLException twice
+            cache = ServerMetadataCache.getInstance(config);
+            ServerMetadataCache spyCache = Mockito.spy(cache);
+            SQLException e = new SQLException("FAIL");
+            Mockito.doThrow(e).doThrow(e).when(spyCache)
+                    .getLastDDLTimestampForTable(any(), any(), eq(Bytes.toBytes(tableName)));
+            ServerMetadataCache.setInstance(spyCache);
+
+            // query using client-2 should fail
+            query(conn2, tableName);
+            Assert.fail("Query should have thrown Exception");
+        }
+        catch (Exception e) {
+            Assert.assertTrue("SQLException was not thrown when last ddl timestamp validation encountered errors twice.", e instanceof SQLException);
+        }
+    }
+  
+    
+    /**
+     * Client-1 creates a table, 2 level of views on it and alters the first level view.
+     * Client-2 queries the second level view, verify that there were 3 cache updates in client-2,
+     * one each for the two views and base table.
+     */
+    @Test
+    public void testSelectQueryOnView() throws Exception {
+        Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+        String url1 = QueryUtil.getConnectionUrl(props, config, "client1");
+        String url2 = QueryUtil.getConnectionUrl(props, config, "client2");
+        String tableName = generateUniqueName();
+        ConnectionQueryServices spyCqs1 = Mockito.spy(driver.getConnectionQueryServices(url1, props));
+        ConnectionQueryServices spyCqs2 = Mockito.spy(driver.getConnectionQueryServices(url2, props));
+        int expectedNumCacheUpdates;
+
+        try (Connection conn1 = spyCqs1.connect(url1, props);
+             Connection conn2 = spyCqs2.connect(url2, props)) {
+
+            // create table using client-1
+            createTable(conn1, tableName, NEVER);
+            upsert(conn1, tableName);
+
+            // create 2 level of views using client-1
+            String view1 = generateUniqueName();
+            String view2 = generateUniqueName();
+            createView(conn1, tableName, view1);
+            createView(conn1, view1, view2);
+
+            // query second level view using client-2
+            query(conn2, view2);
+            expectedNumCacheUpdates = 3; // table, view1, view2
+            Mockito.verify(spyCqs2, Mockito.times(expectedNumCacheUpdates))
+                    .addTable(any(PTable.class), anyLong());
+
+            // alter first level view using client-1 to update its last ddl timestamp
+            alterViewAddColumn(conn1, view1, "foo");
+
+            // invalidate region server cache
+            // TODO: remove this call after PHOENIX-6968 is committed.
+            ServerMetadataCache.resetCache();
+
+            // reset the spy CQSI object
+            Mockito.reset(spyCqs2);
```

Comment:
Normally it is advised never to use mockito.reset ever in tests. 
See [here](https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#resetting_mocks) 
But I think we can use it here.

Reply:
I see, did not know that. My initial approach without reset was a better way then, I can revert to that.

Reply:
But then we won't be able to test for which tables the updateCache was called. I think using Mockito.reset is fine since it is improving our test coverage.